# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/prim_store/backdrop.rs
# Commit: 827ff7c5f5f4
# Full Hash: 827ff7c5f5f4f74aad42f25ccae7b314369c0b11
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2022-05-25 03:42:54
# Regressor Bug: 1769963
# File Overlap Count: 2
# Description:
#   Bug 1769963 - Fix backdrop-filter inside complex transform with CSS filter chain r=gfx-reviewers,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D147124
# ==============================================================================

diff -r aac2eb307b0b -r 827ff7c5f5f4 gfx/wr/webrender/src/prim_store/backdrop.rs
--- a/gfx/wr/webrender/src/prim_store/backdrop.rs	Tue May 24 19:29:01 2022 +0000
+++ b/gfx/wr/webrender/src/prim_store/backdrop.rs	Tue May 24 19:30:19 2022 +0000
@@ -7,89 +7,169 @@
 use crate::internal_types::LayoutPrimitiveInfo;
 use crate::prim_store::{
     InternablePrimitive, PrimitiveInstanceKind, PrimKey, PrimTemplate,
-    PrimTemplateCommonData, PrimitiveStore,
+    PrimTemplateCommonData, PrimitiveStore, PictureIndex,
 };
 use crate::scene_building::IsVisible;
 
 #[cfg_attr(feature = "capture", derive(Serialize))]
 #[cfg_attr(feature = "replay", derive(Deserialize))]
 #[derive(Debug, Clone, Eq, PartialEq, MallocSizeOf, Hash)]
-pub struct Backdrop {
+pub struct BackdropCapture {
+}
+
+#[cfg_attr(feature = "capture", derive(Serialize))]
+#[cfg_attr(feature = "replay", derive(Deserialize))]
+#[derive(Debug, Clone, Eq, PartialEq, MallocSizeOf, Hash)]
+pub struct BackdropRender {
 }
 
-impl From<Backdrop> for BackdropData {
-    fn from(_backdrop: Backdrop) -> Self {
-        BackdropData {
+impl From<BackdropCapture> for BackdropCaptureData {
+    fn from(_backdrop: BackdropCapture) -> Self {
+        BackdropCaptureData {
+        }
+    }
+}
+
+impl From<BackdropRender> for BackdropRenderData {
+    fn from(_backdrop: BackdropRender) -> Self {
+        BackdropRenderData {
         }
     }
 }
 
-pub type BackdropKey = PrimKey<Backdrop>;
+pub type BackdropCaptureKey = PrimKey<BackdropCapture>;
+pub type BackdropRenderKey = PrimKey<BackdropRender>;
 
-impl BackdropKey {
+impl BackdropCaptureKey {
     pub fn new(
         info: &LayoutPrimitiveInfo,
-        backdrop: Backdrop,
+        backdrop_capture: BackdropCapture,
     ) -> Self {
-        BackdropKey {
+        BackdropCaptureKey {
             common: info.into(),
-            kind: backdrop,
+            kind: backdrop_capture,
         }
     }
 }
 
-impl InternDebug for BackdropKey {}
+impl BackdropRenderKey {
+    pub fn new(
+        info: &LayoutPrimitiveInfo,
+        backdrop_render: BackdropRender,
+    ) -> Self {
+        BackdropRenderKey {
+            common: info.into(),
+            kind: backdrop_render,
+        }
+    }
+}
+
+impl InternDebug for BackdropCaptureKey {}
+impl InternDebug for BackdropRenderKey {}
 
 #[cfg_attr(feature = "capture", derive(Serialize))]
 #[cfg_attr(feature = "replay", derive(Deserialize))]
 #[derive(Debug, MallocSizeOf)]
-pub struct BackdropData {
+pub struct BackdropCaptureData {
+}
+
+#[cfg_attr(feature = "capture", derive(Serialize))]
+#[cfg_attr(feature = "replay", derive(Deserialize))]
+#[derive(Debug, MallocSizeOf)]
+pub struct BackdropRenderData {
 }
 
-pub type BackdropTemplate = PrimTemplate<BackdropData>;
+pub type BackdropCaptureTemplate = PrimTemplate<BackdropCaptureData>;
+pub type BackdropRenderTemplate = PrimTemplate<BackdropRenderData>;
 
-impl From<BackdropKey> for BackdropTemplate {
-    fn from(backdrop: BackdropKey) -> Self {
+impl From<BackdropCaptureKey> for BackdropCaptureTemplate {
+    fn from(backdrop: BackdropCaptureKey) -> Self {
         let common = PrimTemplateCommonData::with_key_common(backdrop.common);
 
-        BackdropTemplate {
+        BackdropCaptureTemplate {
             common,
             kind: backdrop.kind.into(),
         }
     }
 }
 
-pub type BackdropDataHandle = InternHandle<Backdrop>;
+impl From<BackdropRenderKey> for BackdropRenderTemplate {
+    fn from(backdrop: BackdropRenderKey) -> Self {
+        let common = PrimTemplateCommonData::with_key_common(backdrop.common);
 
-impl Internable for Backdrop {
-    type Key = BackdropKey;
-    type StoreData = BackdropTemplate;
+        BackdropRenderTemplate {
+            common,
+            kind: backdrop.kind.into(),
+        }
+    }
+}
+
+pub type BackdropCaptureDataHandle = InternHandle<BackdropCapture>;
+pub type BackdropRenderDataHandle = InternHandle<BackdropRender>;
+
+impl Internable for BackdropCapture {
+    type Key = BackdropCaptureKey;
+    type StoreData = BackdropCaptureTemplate;
     type InternData = ();
     const PROFILE_COUNTER: usize = crate::profiler::INTERNED_BACKDROPS;
 }
 
-impl InternablePrimitive for Backdrop {
+impl Internable for BackdropRender {
+    type Key = BackdropRenderKey;
+    type StoreData = BackdropRenderTemplate;
+    type InternData = ();
+    const PROFILE_COUNTER: usize = crate::profiler::INTERNED_BACKDROPS;
+}
+
+impl InternablePrimitive for BackdropCapture {
     fn into_key(
         self,
         info: &LayoutPrimitiveInfo,
-    ) -> BackdropKey {
-        BackdropKey::new(info, self)
+    ) -> BackdropCaptureKey {
+        BackdropCaptureKey::new(info, self)
     }
 
     fn make_instance_kind(
-        _key: BackdropKey,
-        data_handle: BackdropDataHandle,
+        _key: BackdropCaptureKey,
+        data_handle: BackdropCaptureDataHandle,
         _prim_store: &mut PrimitiveStore,
         _reference_frame_relative_offset: LayoutVector2D,
     ) -> PrimitiveInstanceKind {
-        PrimitiveInstanceKind::Backdrop {
+        PrimitiveInstanceKind::BackdropCapture {
             data_handle,
         }
     }
 }
 
-impl IsVisible for Backdrop {
+impl InternablePrimitive for BackdropRender {
+    fn into_key(
+        self,
+        info: &LayoutPrimitiveInfo,
+    ) -> BackdropRenderKey {
+        BackdropRenderKey::new(info, self)
+    }
+
+    fn make_instance_kind(
+        _key: BackdropRenderKey,
+        data_handle: BackdropRenderDataHandle,
+        _prim_store: &mut PrimitiveStore,
+        _reference_frame_relative_offset: LayoutVector2D,
+    ) -> PrimitiveInstanceKind {
+        PrimitiveInstanceKind::BackdropRender {
+            data_handle,
+            pic_index: PictureIndex::INVALID,
+        }
+    }
+}
+
+impl IsVisible for BackdropCapture {
     fn is_visible(&self) -> bool {
         true
     }
 }
+
+impl IsVisible for BackdropRender {
+    fn is_visible(&self) -> bool {
+        true
+    }
+}